File: server/src/graphs/state.js
import { Annotation, MessagesAnnotation } from "@langchain/langgraph";

export const GraphState = Annotation.Root({
  ...MessagesAnnotation.spec,

  userInput: Annotation({
    reducer: (_, next) => next,
    default: () => "",
  }),

  // T16: 安全护栏拦截标记
  guardrailBlocked: Annotation({
    reducer: (_, next) => next,
    default: () => false,
  }),

  // 单意图（保留兼容：取 intents[0]）
  intent: Annotation({
    reducer: (_, next) => next,
    default: () => "",
  }),

  // 多意图数组 — 支持 fan-out 并行执行
  // reducer 使用追加模式，避免并行分支覆盖
  intents: Annotation({
    reducer: (a, b) => [...a, ...b],
    default: () => [],
  }),

  orderResult: Annotation({
    reducer: (_, next) => next,
    default: () => null,
  }),

  ragResult: Annotation({
    reducer: (_, next) => next,
    default: () => "",
  }),

  // 知识库检索来源（供前端参考来源卡片展示）
  ragSources: Annotation({
    reducer: (_, next) => next,
    default: () => [],
  }),

  finalAnswer: Annotation({
    reducer: (_, next) => next,
    default: () => "",
  }),
});

File: backend_py/graphs/state.py
"""
工作流状态定义
对照 JS: Annotation.Root({ ...MessagesAnnotation.spec, userInput: Annotation({...}), ... })
Python:  TypedDict + Annotated[类型, reducer函数]
"""
from typing import TypedDict, Annotated, Optional
from langgraph.graph.message import add_messages


class GraphState(TypedDict):
    # 继承内置的消息列表管理（add_messages reducer 会自动追加消息）
    messages: Annotated[list, add_messages]

    # 当前用户输入
    userInput: str

    # 意图识别结果：order | knowledge | general
    intent: str

    # 订单 Agent 结果
    orderResult: Optional[dict]

    # RAG 检索结果
    ragResult: str

    # 最终回答
    finalAnswer: str